File: 프로그래머스/lv3/87694. 아이템 줍기/아이템 줍기.js
function solution(rectangle, characterX, characterY, itemX, itemY) {
    let answer = 0

        const map = Array.from({length:102},()=>Array.from({length:102},()=>-1))
    
    rectangle.forEach((r)=>{
        const [x1,y1,x2,y2] = r.map((el)=>el*2)
        
        for(let i=x1; i<=x2; i++) { //2 14 
            for(let j=y1; j<=y2; j++) { // 2 8
                if(x1<i && i<x2 && y1<j && j<y2) { 
                    map[i][j] = 0
                } else if(map[i][j] != 0) { 
                    map[i][j] = 1 
                } 
            }
        }
    })

    const dx = [-1,1,0,0]
    const dy = [0,0,-1,1]
    
    let q = []
    q.push([characterX*2, characterY*2])
        const visited = Array.from({length:102},()=>Array.from({length:102},()=>1))
     visited[characterX*2][characterY*2] = 0 // 시작 지점 0으로 초기화
    
    while(q.length) {
        let [x,y] = q.shift()
        if((x === (itemX * 2)) && (y === (itemY * 2))) {   
            answer = parseInt(visited[x][y] / 2)  // 아이템 획득
            break 
        }
        
        // 상하좌우 탐색
        for(let i=0;i<4;i++) {
            let nx = x + dx[i] // 2 + (-1) => 1
            let ny = y + dy[i] // 6 + 0 => 6

            if(map[nx][ny] === 1 && visited[nx][ny] === 1) { // 테두리이고 방문하지 않았으면
                q.push([nx,ny]) // 큐에 추가
                visited[nx][ny] = visited[x][y]+1
            }
        }
    }
    return answer
}

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangle, characterX, characterY, itemX, itemY):
    answer = 0
    
    MAX = 102  # 50*2+2
    map = [[-1] * MAX for _ in range(MAX)]
    
    for r in rectangle:
        # 대각선으로 맞닿은 사각형에서 이동 못하게 두 배로 늘림
        x1,y1,x2,y2 = [i *2 for i in r] 
               
        for i in range(x1,x2+1):
            for j in range(y1,y2+1):
                # 내부 0으로 채우기
                if x1<i<x2 and y1<j<y2:
                    map[i][j] = 0
                # 내부X + 0이 아님 => 테두리
                elif map[i][j] != 0:
                    map[i][j] = 1
        # print(map)
                    
    dx,dy = [-1,1,0,0], [0,0,-1,1]
        
    q = deque()
    # 시작 위치 지정
    q.append((characterX*2,characterY*2))
    
    visited = [[-1] * MAX for _ in range(MAX)]
    visited[characterX*2][characterY*2] = 0 # 시작점 거리
    
    while q:
        x,y = q.popleft()
        
        # 도착
        if x == itemX*2 and y == itemY*2:
            return visited[x][y] // 2
        
        # 상하좌우 탐색
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            
            if 0 <= nx < MAX and 0 <= ny < MAX: # 맵 내부
                # 테두리 + 미방문
                if map[nx][ny] == 1 and visited[nx][ny]== -1:
                    visited[nx][ny] = visited[x][y] + 1 # 길이 저장
                    q.append((nx,ny))
